graviton/dryrun.py

```diff
@@ -0,0 +1,919 @@
+    result = None
+    if not ok:
+        result = f"{msg}: {error}" + _msg_if(msg)
+        if enforce:
+            assert not error, result
+
+    return ok, result
+
+
+def assert_global_state_contains(delta_value, txn_index, txns_res, msg=None):
+    if txn_index is not None and (txn_index < 0 or txn_index >= len(txns_res)):
+        _fail(f"txn index {txn_index} is out of range [0, {len(txns_res)})")
+
+    found = False
+    all_global_deltas = []
+    for idx, txn_res in enumerate(txns_res):
+        # skip if txn_index is set
+        if txn_index is not None and idx != txn_index:
+            continue
+        if (
+            "global-delta" in txn_res
+            and txn_res["global-delta"] is not None
+            and len(txn_res["global-delta"]) > 0
+        ):
+            found = DryRunHelper.find_delta_value(txn_res["global-delta"], delta_value)
+            if not found and idx == txn_index:
+                msg = (
+                    msg
+                    if msg is not None
+                    else f"{delta_value} not found in {txn_res['global-delta']}"
+                )
+                _fail(msg)
+            if found:
+                break
+            all_global_deltas.extend(txn_res["global-delta"])
+        elif idx == txn_index:
+            _fail("no global state from dryrun")
+
+    if not found:
+        msg = (
+            msg
+            if msg is not None
+            else f"{delta_value} not found in any of {all_global_deltas}"
+        )
+        _fail(msg)
+
+
+def assert_local_state_contains(addr, delta_value, txn_index, txns_res, msg=None):
+    if txn_index is not None and (txn_index < 0 or txn_index >= len(txns_res)):
+        _fail(f"txn index {txn_index} is out of range [0, {len(txns_res)})")
+
+    found = False
+    all_local_deltas = []
+    for idx, txn_res in enumerate(txns_res):
+        # skip if txn_index is set
+        if txn_index is not None and idx != txn_index:
+            continue
+        if (
+            "local-deltas" in txn_res
+            and txn_res["local-deltas"] is not None
+            and len(txn_res["local-deltas"]) > 0
+        ):
+            for local_delta in txn_res["local-deltas"]:
+                addr_found = False
+                if local_delta["address"] == addr:
+                    addr_found = True
+                    found = DryRunHelper.find_delta_value(
+                        local_delta["delta"], delta_value
+                    )
+                    if not found and idx == txn_index:
+                        msg = (
+                            msg
+                            if msg is not None
+                            else f"{delta_value} not found in {local_delta['delta']}"
+                        )
+                        _fail(msg)
+                    if found:
+                        break
+                    all_local_deltas.extend(local_delta["delta"])
+            if not addr_found and idx == txn_index:
```

Comment:
@tzaffi Sanity check:  Based on indents, it seems `add_found` is _not_ lexically available.  Can you double check me?

As I see it, the highlighted statement is _outside_ the for-loop.  If that's accurate, then I'm pointing out the concern because the check won't work as intended.

Reply:
Good catch.

Inheriting old bad code here from pysdk. I'll correct and let the integration tests be the judge.